src/tasks/tasks.ts

```diff
@@ -300,11 +317,84 @@ export class Tasks {
 
     /**
      * Apply new captcha solutions to captcha dataset and recalculate merkle tree
-     * @param {string} datasetId
      */
-    async calculateCaptchaSolutions (datasetId: string) {
-        // TODO run this on a predefined schedule as updating the dataset requires committing an updated
-        // captcha_dataset_id to the blockchain
+    async calculateCaptchaSolutions () {
+        try {
+            const captchaFilePath = this.captchaSolutionConfig.captchaFilePath
+            const currentDataset = parseCaptchaDataset(loadJSONFile(captchaFilePath) as JSON)
+            if (!currentDataset.datasetId) {
+                return
+            }
+            const unsolvedCaptchas = await this.db.getAllCaptchasByDatasetId(currentDataset.datasetId as string, CaptchaStates.Unsolved)
+
+            const totalNumberOfSolutions = this.captchaSolutionConfig.requiredNumberOfSolutions
+            const winningPercentage = this.captchaSolutionConfig.solutionWinningPercentage
+            const winningNumberOfSolutions = Math.round(totalNumberOfSolutions * (winningPercentage / 100))
+            let solutionsToUpdate: CaptchaSolutionToUpdate[] = []
+
+            if (unsolvedCaptchas && unsolvedCaptchas.length > 0) {
+                for (let i = 0; i < unsolvedCaptchas.length; i++) {
+                    const solutions = await this.db.getAllSolutions(unsolvedCaptchas[i].captchaId)
+                    if (solutions && solutions.length >= totalNumberOfSolutions) {
+                        const solutionsWithCount = {}
+                        for (let i = 0; i < solutions.length; i++) {
+                            const previousCount = solutionsWithCount[JSON.stringify(solutions[i].solution)]?.solutionCount || 0
+                            solutionsWithCount[JSON.stringify(solutions[i].solution)] = {
+                                captchaId: solutions[i].captchaId,
+                                solution: solutions[i].solution,
+                                salt: solutions[i].salt,
+                                solutionCount: previousCount + 1
+                            }
+                        }
+                        solutionsToUpdate = solutionsToUpdate.concat(
+                            Object.values(solutionsWithCount)
+                                .filter(({ solutionCount }: any) => solutionCount >= winningNumberOfSolutions)
+                                .map(({ solutionCount, ...otherAttributes }: any) => otherAttributes))
+                    }
+                    if (solutionsToUpdate.length > 0 && i === unsolvedCaptchas.length - 1) {
```

Comment:
from the condition it looks like this would need to happen outside of the loop?

Reply:
Yes just to make sure the logic runs after the loop completes. Updated it put this block outside the loop since no async task is there in the loop.